messaging/src/main/scala/uk/ac/wellcome/messaging/worker/Worker.scala

```diff
@@ -0,0 +1,64 @@
+package uk.ac.wellcome.messaging.worker
+
+import java.time.Instant
+
+import uk.ac.wellcome.messaging.worker.monitoring.{MonitoringClient, ProcessMonitor, SummaryRecorder}
+import uk.ac.wellcome.messaging.worker.result.models.{DeterministicFailure, PostProcessFailure}
+import uk.ac.wellcome.messaging.worker.result.Result
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait Worker[
+ProcessMonitoringClient <: MonitoringClient,
+Work,
+Summary,
+WorkerMessage,
+Action,
+MessageProcess <: WorkerProcess[
+  Work,
+  Summary]
+] extends SummaryRecorder
+  with ProcessMonitor[ProcessMonitoringClient] {
+
+  protected val process: MessageProcess
+
+  protected def toWork(message: WorkerMessage): Future[Work]
+
+  protected def toMessageAction(result: Result[_]): Future[Action]
+
+  protected def processMessage(
+                                id: String,
+                                message: WorkerMessage
+                              )(implicit
+                                monitoringClient: ProcessMonitoringClient,
+                                ec: ExecutionContext
+
+                              ): Future[(WorkerMessage, Action)] = {
+    val startTime = Instant.now
+
+    val result = for {
+      work <- toWork(message)
+      result <- process.run(work)
+    } yield result
+
+    val recoveredResult = result.recover {
+      case e => DeterministicFailure(id, e)
```

Comment:
So to be clear: it’s the responsibility of the caller who defines `process.run()` to determine if a failure is non-deterministic? 👍

Reply:
Yes.